### PR TITLE
SR-MPLS: extending API to support IP-in-MPLS tunneling

### PR DESCRIPTION
**Pull request by phshaikh**
Proposing API extensions to be able to use MPLS for Segment Routing L3 traffic.

### CODE REVIEW

**Comment by kcudnik** [comment]
Please correct all warnings and errors: https://sonic-jenkins.westus2.cloudapp.azure.com/job/sai-meta-pr-build/1929/console
i also think this PR will require some changes to parser since new type label was used, i will take a look after errors will be corrected

**Comment by phshaikh** [comment]
@kcudnik Thank you for response. Are there set of commands shall I run before I resubmit changes for review? 

I did following
# cd meta
# make

And it succeeded, I am trying to see if there're more commands I can run locally before sending updates

**Comment by kcudnik** [comment]
that everything you can run locally, and if succeeded then you are good to go

File: inc/saimpls.h [hunk 1]
```diff
@@ -81,6 +81,31 @@ typedef enum _sai_inseg_entry_attr_t
...
 
+    /**
+     * @brief Use TC for a label.
+     *
+     * Associate TC using SAI_SWITCH_ATTR_QOS_EXP_TO_TC_MAP or SAI_PORT_ATTR_QOS_EXP_TO_TC_MAP.
+     * Mainly used for E-LSP kind of MPLS tunnels, where EXP infers both TC and COLOR
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
```
Comment (lguohan) [inline inc/saimpls.h]: is this backward compatible? should true to be default behavior or false?

File: inc/sainexthop.h [hunk 2]
```diff
@@ -229,6 +229,28 @@ typedef enum _sai_next_hop_attr_t
...
+    /**
+     * @brief Encap tunnel-id for MPLS PUSH
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @objects SAI_OBJECT_TYPE_TUNNEL
+     * @allownull true
+     * @condition SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_MPLS
+     */
+    SAI_NEXT_HOP_ATTR_MPLS_ENCAP_TUNNEL_ID,
```
Comment (lguohan) [inline inc/sainexthop.h]: is this behavior backward compatible? it looks now if the next hop is mpls, then user must include this new attribute? it seems like a non-backward compatible behavior.

File: inc/sainexthop.h [hunk 3]
```diff
@@ -229,6 +229,28 @@ typedef enum _sai_next_hop_attr_t
...
+    /**
+     * @brief Decap tunnel-id for MPLS PHP
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @objects SAI_OBJECT_TYPE_TUNNEL
+     * @allownull true
+     * @condition SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_IP
+     */
+    SAI_NEXT_HOP_ATTR_MPLS_DECAP_TUNNEL_ID,
```
Comment (lguohan) [inline inc/sainexthop.h]: is this behavior backward compatible? it looks now if the next hop is IP, then user must include this new attribute? it seems like a non-backward compatible behavior.

Comment (phshaikh) [reply inc/sainexthop.h]: For decapsulation (PHP) this new optional attribute is proposed. If provided, then tunnel modes can be used to identify TTL and QoS treatment for outgoing traffic. 

Next hop of type IP is used for PHP as per https://github.com/opencomputeproject/SAI/blob/master/doc/MPLS/SAI-Proposal-MPLS-ver4.docx

File: inc/sainexthop.h (hunk 2 above)
Comment (phshaikh) [reply inc/sainexthop.h]: this is not mandatory attribute, it is proposed for PUSH operation to indicate TTL and QoS mode

File: inc/saimpls.h (hunk 1 above)
Comment (phshaikh) [reply inc/saimpls.h]: There was no QoS mapping in existing SAI spec for MPLS traffic. 
Default behavior proposal is to use "QoS map", i think comment is wrong here (Use TC for a label should be Use TC from a QoS Map).

**Comment by phshaikh** [comment]
Some comments in the last meeting
- split this pull request in to smaller pull requests
- move decap tunnel id into inseg entry (by Haidong) so same next hop can be reused for tunnel termination at PHP nodes

File: inc/sainexthop.h (hunk 3 above)
Comment (phshaikh) [reply inc/sainexthop.h]: I plan to move this to InSegEntry as per comment by with Haidong last week. This is to be used in conjunction with IP next hop to provide decapsulation behavior for PHP

**Comment by phshaikh** [comment]
Closing this, divided this pull request into many smaller pull requests - 
1) https://github.com/opencomputeproject/SAI/pull/1058
2) https://github.com/opencomputeproject/SAI/pull/1060
3) https://github.com/opencomputeproject/SAI/pull/1062
4) https://github.com/opencomputeproject/SAI/pull/1064
5) https://github.com/opencomputeproject/SAI/pull/1066

**Pull request closed without merging**